### PR TITLE
feat: App::built_command

### PR DESCRIPTION
**Pull request by scooter-dangle**
Who needs tests????

- - -

~Need to add tests, but will take some time to get around to that. Might still be productive to review in the intervening time.~

Added tests!

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -216,93 +224,111 @@ fn complete_command(
...
-//
-//     output_complete_command(escaped_cmd, fish_options, truncated_description, built_command_output)
+const MAX_DESCRIPTION_WIDTH: usize = 78;
+const TRUNCATION_SUFFIX: &'static str = "…";
 
-// TODO args / arg types?
-fn built_command() {
-    unimplemented!()
+fn last_char_index(string: &str) -> Option<usize> {
+    string.rfind(|_| true)
```
Comment (dlrobertson) [inline src/main.rs]: Is there a reason not to use `str::len`?

Comment (scooter-dangle) [reply src/main.rs]: The last byte index might not not be on a UTF-8 character boundary.

Comment (dlrobertson) [reply src/main.rs]: Ah, good to know.

Comment (scooter-dangle) [reply src/main.rs]: Ended up not needing this. Misinterpreted the original Python's description truncation code.

File: Cargo.toml [hunk 2]
```diff
@@ -10,3 +10,6 @@ itertools = "*"
 regex = "*"
 lazy_static = "*"
 bstr = "*"
+
+[dev-dependencies]
+pretty_assertions = "*"
```
Comment (justinaustin) [inline Cargo.toml]: unrelated to the PR, but will this eventually end up on crates.io? And if so do they allow wildcard dependency versions?

Comment (scooter-dangle) [reply Cargo.toml]: Hmm. Good question. Initially we'll probably only be distributing a binary, which I don't have experience with on crates.io. Not sure if it gives allowances to binaries on the manifest if they also contain a lockfile.

Comment (PureW) [reply Cargo.toml]: Assuming the library uses semantic versioning, I usually wildcard the patch-version, and lock the major and minor version. In this case, it's a dev-dependency for the unittests which decreases the risk of a complete wildcard.

Important in this context is whether you check in the `Cargo.toml` file (recommended for binaries) or not (recommended for libraries). If you don't check in the lock-file, you have a higher risk of build-failures due to changes in external dependencies.

File: src/main.rs [hunk 3]
```diff
@@ -194,30 +209,213 @@ fn test_lossy_unicode() {
...
+            (if option.len() == 2 { "s" } else { "o" }, 1)
+        } else {
+            continue;
+        };
+
+        let option = format!(
+            "-{} {}",
+            fish_opt,
+            // Direct indexing of `option` won't panic due to how `num_dashes`
+            // is calculated.
```
Comment (PureW) [inline src/main.rs]: You should add `(I promise)`.

File: src/main.rs [hunk 4]
```diff
@@ -194,30 +209,213 @@ fn test_lossy_unicode() {
...
+            out.push(option);
+        }
+    }
+
+    out
+}
+
+#[test]
+fn test_fish_options() {
+    let output: Vec<String> = vec!["-s 'f'".into(), "-l 'force'".into()];
```
Comment (PureW) [inline src/main.rs]: Tiny nit, it's usually helpful to casual readers to name the _true_ answer to something more descriptive like `expected_output` or similar.

File: src/main.rs [hunk 5]
```diff
@@ -194,30 +209,213 @@ fn test_lossy_unicode() {
...
+    let mut existing_options: HashSet<String> = Default::default();
+    assert_eq!(fish_options(options, &mut existing_options), output);
+    assert_eq!(
+        output.iter().cloned().collect::<HashSet<_>>(),
+        existing_options
+    );
+
+    use std::iter::once;
+
+    let output: Vec<String> = vec!["-l 'force'".into()];
```
Comment (PureW) [inline src/main.rs]: Another tiny tiny nit: Perhaps enclose line 277-298 and 266-273 in braces to lower risk of stray variables doing naughty things?

File: src/main.rs [hunk 6]
```diff
@@ -194,30 +209,213 @@ fn test_lossy_unicode() {
...
+            .chain(once("-l 'something'".to_string()))
+            .collect::<HashSet<_>>(),
+        existing_options,
+    );
+}
+
+/// # Panics
+/// If `max_length` is zero.
+fn char_truncate_string(string: &str, max_length: usize, truncator: char) -> Cow<str> {
+    let mut chars = string
```
Comment (PureW) [inline src/main.rs]: `s/chars/char_indices/`?

**Comment by scooter-dangle** [comment]
Thanks for the review, @justinaustin! Good to go, @PureW!

**Pull request merged**